### PR TITLE
uart: ns16550: support Apollo Lake PRV_CLOCK_PARAMS (PCP) register

### PR DESCRIPTION
**Pull request by ghost**
The UARTs on the Apollo Lake SoCs have PLLs that feed the baud rate
generators. This patch allows a user to specify custom M/N values for
those PLLs when custom/high-speed baud rates are required.

I'm not entirely satisfied with the way the PCP values are configured,
because it requires tweaking data in both Kconfig and DeviceTree. For
the time being I've merely taken my cue from another similar feature
(the DLF register support) and have punted on figuring out the "right
way" to expose UART configuration to the application.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

### CODE REVIEW

File: drivers/serial/Kconfig.ns16550 [hunk 1]
```diff
@@ -25,6 +25,16 @@ config UART_NS16550_DLF
 
 	  Says n if you are not sure if hardware supports this.
 
+config UART_NS16550_PCP
+	bool "Enable Apollo Lake PRV_CLOCK_PARAMS (PCP) support"
```
Comment (nashif) [inline drivers/serial/Kconfig.ns16550]: is this really an exclusive feature of apollo lake or is that a feature of the NS116500 used on that platform. Lets try and keep the drivers agnostic to the platform they are being used on.

**Review by nashif — DISMISSED** [review]

File: drivers/serial/Kconfig.ns16550 (hunk 1 above)
Comment (gnuless) [reply drivers/serial/Kconfig.ns16550]: As far as I know, this is specific to the Apollo Lake's implementation of the NS16550.  Unfortunately this configuration isn't board- or SoC-level; rather each UART has its own clock block which is set by manipulating device registers, which requires knowledge of where and how the device is mapped, and so the driver is IMO "best" place to put this.

I agree that this pollution of the driver is suboptimal, but it does follow the example set for the DLF register, which is specific to the NS16550 implementation found on some of the Quarks.

Comment (dcpleung) [reply drivers/serial/Kconfig.ns16550]: DLF is a Synopsys addition to the IP block which applies to more than one SoC. For this, I wonder if we should set up the clock in early SoC init code.

**Comment by gnuless** [comment]
How would you propose going about that?  I'm supposed to enumerate the PCI
bus to get the serial devices and set the registers there outside the
driver?

**Review by tyrantlizard77 — APPROVED** [review]
Just saw this patch. My understanding of serial is a bit limited but hopefully this helps the discussion.

> How would you propose going about that?

So the way SOC-specific initialization is typically done is to implement a SYS_INIT() function in the soc-specific C file, which in the case of Apollo Lake would be an addition to soc/x86/apollo_lake/soc.c roughly like this:

```
static int soc_apollo_lake_init(struct device *arg)
{
  ARG_UNUSED(arg);

  ... do stuff ...
}
SYS_INIT(soc_apollo_lake, PRE_KERNEL_1, 0);
```

> I'm supposed to enumerate the PCI bus to get the serial devices and set the registers there outside the driver?

One one hand, it *appears* we know at build time what serial devices we are dealing with, since we instantiate uart_ns16550_device_config uart_ns16550_dev_cfg_1 ... 3 based on whether CONFIG_UART_NS16550_PORT_0 .. 3 are defined. We also know the port addresses for all of these UARTS since they come from DTS and not at runtime ( DT_UART_NS16550_PORT_0_BASE_ADDR ... etc). So it looks possible, in a soc_apollo_lake_init() function, to poke the PCP for all the enabled UARTS outside of the main driver.

However I'm seeing in the ns16550 init function that, before we poke any registers, that ns16550_pci_uart_scan() is called to determine if the device actually exists, which results in the init function failing with -ENOTSUP before anything actually happens if not. So it may be unsafe to try to poke this stuff in the SOC code without repeating that test there, in which case it doesn't seem worth the hassle to move this there to begin with.

So the hundred dollar question is: do we really know the UART configuration entirely at build time? If I'm understanding @gnuless correctly, we do not.

Zephyr, to date, has been targeting devices where the entire hardware configuration is known at build time, either through Kconfig or DTS defines. In a world where we can't know for sure whether the device is really there, I think the current approach of doing this in the driver is appropriate.

**Review by dcpleung — APPROVED** [review]
We probably need a way to control clocks from the drivers, and maybe better ways to handle quirks on the driver side. Lacking these, this patch looks fine. A future enhancement would be to calculate the M/N values based on the clock frequency from DTS.

**Comment by gnuless** [comment]
> A future enhancement would be to calculate the M/N values based on the clock frequency from DTS.

For any given PLL, not all values of M and N will work, and since we're talking about a ratio here, many choices of M and N exist for any given requirement.  This could also potentially require a DTS configuration to specify the input clock to the UART block, which is 100MHz on the UP^2 but could (potentially, I don't know) be different on different boards.  This is all above and beyond the purview of the build system, especially as this is, at present, for specialized applications and those responsible for said applications will presumably know better than we do what M and N values they want.

**Comment by gnuless** [comment]
I appreciate the time you took to answer these questions, but they were meant rhetorically.

**Comment by gnuless** [comment]
> Just saw this patch. My understanding of serial is a bit limited but hopefully this helps the discussion.

I'm sorry you took the time out of your day to look into this.  These were rhetorical questions.

**Pull request merged**